util/include/util/threadpool.h

```diff
@@ -0,0 +1,308 @@
+   *
+   * The single-threaded "pool" synchronously runs the job immediately.
+   *
+   * @param jobInfo  The job object to work on.
+   */
+  void enqueue(JobData jobInfo)
+  {
+    func(jobInfo);
+  }
+
+  /**
+   * @brief Has no effect in single-threaded operation as enqueue()
+   * automatically runs the job function.
+   */
+  void nudge()
+  {}
+
+  /**
+   * @brief Has no effect in single-threaded operation as enqueue()
+   * automatically runs the job function.
+   */
+  void wait()
+  {}
+
+private:
+  /**
+   * The function which is executed on incoming jobs.
+   */
+  Function func;
+};
+
+/**
+ * @brief A simple thread pool which iterates a set of jobs dynamically.
+ *
+ * This class creates N worker threads in the background which are waken up
+ * as jobs are added to the queue. Each worker takes a single job and executes
+ * it, and the threads return to sleep.
+ *
+ * @tparam JobData   Jobs are represented in a custom, user-defined structure.
+ * @tparam Function  A user defined functor which the workers call to do the
+ * actual work. This functor must accept a JobData as its argument.
+ */
+template <typename JobData,
+          typename Function = std::function<void (JobData)>>
+class JobQueueThreadPool
+  : public JobQueueThreadPoolBase<JobData>
+{
+public:
+  /**
+   * Create a new thread pool with the given number of threads and using the
+   * given function as its work logic.
+   *
+   * @param threadCount  The number of worker threads to create.
+   * @param func         The function to execute on the enqueued jobs.
+   */
+  JobQueueThreadPool(size_t threadCount, Function func)
+    : threadCount(threadCount), _die(false)
+  {
+    for (size_t i = 0; i < threadCount; ++i)
+    {
+      _threads.emplace_back(std::thread(
+        &JobQueueThreadPool<JobData, Function>::worker,
+        this, func));
+    }
+  }
+
+  ~JobQueueThreadPool()
+  {
+    if (!_die)
+      wait();
+  }
+
+  /**
+   * @brief Enqueue a new job to be executed by the thread pool.
+   *
+   * @warning Job execution might start immediately at enqueue's return!
+   *
+   * @param jobInfo  The job object to work on.
+   */
+  void enqueue(JobData jobInfo)
+  {
+    {
+      std::lock_guard<std::mutex> lock(_lock);
+      _queue.push(jobInfo);
+    }
+
+    _signal.notify_all();
+  }
+
+  /**
+   * @brief This method manually generates a signal to wake up the workers.
+   *
+   * @note Sometimes due to optimisation reasons, all enqueue() calls can
+   * return earlier than the worker threads set up and start waiting for jobs.
+   */
+  void nudge()
+  {
+    _signal.notify_all();
+  }
+
+  /**
+   * @brief Notify all workers to exit after doing the remaining work
+   * and wait for the threads to die.
+   */
+  void wait()
+  {
+    _die = true;
+    _signal.notify_all();
+
+    for (std::thread& t : _threads)
+    {
+      // Keep nudging the threads so that the t.join() don't grind the
+      // system to a halt.
+      _signal.notify_all();
```

Comment:
Why do we call notify_all here? Could you sketch the scenario when the omission will result in a halt?